### PR TITLE
[SE-0155] Default Arguments in Enum Cases

### PR DESCRIPTION
**Pull request by CodaFi**
Lift the restriction on default arguments in enum cases.  This
necessitated a number of odd tradeoffs stemming mostly from the
fact that enum cases are not a form of decl context.

An entity has been added to the mangling for default arguments that
mangle the enum case in addition to the entity's decl context.

SILGenExpr has regrown the ability to emit tuple shuffles that involve
caller and callee-side default arguments to support the open-coding path
in SILGenApply.  This could probably be made to use the machinery there
to cough up an RValue that we could use as the payload, but that
involves some deep refactoring.

All in all, this approach seems like the least evil of a whole lot of evil things.  I'm definitely open to better ways to do all of this.

### CODE REVIEW

**Comment by rjmccall** [comment]
I'll review this, but it probably needs an evolution proposal.

**Comment by CodaFi** [comment]
@rjmccall This is a part of [0155](https://github.com/apple/swift-evolution/blob/master/proposals/0155-normalize-enum-case-representation.md#default-parameter-values-for-enum-constructors)

**Review by rjmccall — CHANGES_REQUESTED** [review]
I see.  SE-0155 keeps surprising me with how much was packed into it.

File: lib/AST/DeclContext.cpp [hunk 1]
```diff
@@ -287,6 +287,9 @@ ResilienceExpansion DeclContext::getResilienceExpansion() const {
     // Default argument initializer contexts have their resilience expansion
     // set when they're type checked.
     if (isa<DefaultArgumentInitializer>(dc)) {
+      if (auto *ED = dyn_cast<EnumDecl>(dc->getParent())) {
```
[the remark this answers is not in the document]
Comment (CodaFi) [reply lib/AST/DeclContext.cpp]: This sounds like a plan. I’ll do the plumbing in a separate patch.

**Pull request closed without merging**